pkg/exporter/grpc_packets.go

```diff
@@ -0,0 +1,78 @@
+package exporter
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gopacket"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/netobserv/netobserv-ebpf-agent/pkg/flow"
+	grpc "github.com/netobserv/netobserv-ebpf-agent/pkg/grpc/packet"
+	"github.com/netobserv/netobserv-ebpf-agent/pkg/pbpacket"
+)
+
+type GRPCPacketProto struct {
+	hostIP     string
+	hostPort   int
+	clientConn *grpc.ClientConnection
+}
+
+var gplog = logrus.WithField("component", "packet/GRPCPackets")
+
+// WritePacket writes the given packet data out to the file.
+func writeGRPCPacket(ci gopacket.CaptureInfo, data []byte, conn *grpc.ClientConnection) error {
+	if ci.CaptureLength != len(data) {
+		return fmt.Errorf("capture length %d does not match data length %d", ci.CaptureLength, len(data))
+	}
+	if ci.CaptureLength > ci.Length {
+		return fmt.Errorf("invalid capture info %+v:  capture length > length", ci)
+	}
+	gplog.Debugf("Sending Packet to client. Length: %d", len(data))
+	b, err := GetPacketHeader(ci)
+	if err != nil {
+		return fmt.Errorf("error writing packet header: %w", err)
+	}
+	// write 16 byte packet header & data all at once
+	_, err = conn.Client().Send(context.TODO(), &pbpacket.Packet{
+		Pcap: &anypb.Any{
+			Value: append(b, data...),
+		},
+	})
+	return err
+}
+
+func StartGRPCPacketSend(hostIP string, hostPort int) (*GRPCPacketProto, error) {
+	clientConn, err := grpc.ConnectClient(hostIP, hostPort)
+	if err != nil {
+		return nil, err
+	}
+	return &GRPCPacketProto{
+		hostIP:     hostIP,
+		hostPort:   hostPort,
+		clientConn: clientConn,
+	}, nil
+}
+
+func (p *GRPCPacketProto) ExportGRPCPackets(in <-chan []*flow.PacketRecord) {
+	for packetRecord := range in {
+		if len(packetRecord) > 0 {
+			for _, packet := range packetRecord {
+				packetStream := packet.Stream
+				packetTimestamp := packet.Time
+				if len(packetStream) != 0 {
+					captureInfo := gopacket.CaptureInfo{
+						Timestamp:     packetTimestamp,
+						CaptureLength: len(packetStream),
+						Length:        len(packetStream),
+					}
+					err := writeGRPCPacket(captureInfo, packetStream, p.clientConn)
```

Comment:
I wonder if we can define grpc as array of pcap and write all the record at once instead of writing every packet this will probably be more efficient WDYT?

Reply:
I can give a try indeed. What would be the maximum acceptable total size here ?

A single packet can be huge so I may need to check about the next(s) before sending

Reply:
For the collector side, it doesn't make sense to get half of a packet here. It would require extra work to sync and avoid concurrency

We should keep this simple and write packets entirely; but allowing array of them with a reasonnable upper limit

Reply:
are we ok to limit the packet header size from the ebpf to say 256 bytes max if the packet is fragmented or has encrypted data it has no value copy all of that to user space WDYT ?

Reply:
As we discussed today, it's still interesting to capture the payload for non encrypted packets. Giving the ability to get only the firsts 256 bytes is still interesting if wireshark manages that seamlessly but in that case, flows already cover these informations.